File: Difficulty: Medium/Minimum Sum Subarray/minimum-sum-subarray.java
class Solution {
    public int minSubarraySum(int[] arr) {
        // code here
        // kadens sehua hai
        // int minSum = arr[0],currentSum = 0,n = arr.length;
        // for(int i = 0;i < n;i++){
        //     currentSum += arr[i];
        //     if(currentSum < minSum ){
        //         minSum = currentSum;
        //     }
        //     if(currentSum > 0){
        //         currentSum = 0;
        //     }
        // }
        // return minSum;
        
        // window sliding sa hua hsi
        int minSum = Integer.MAX_VALUE; // Sabse badi value se shuru kiya
        int currentSum = 0;
        int left = 0;                   // Window ka starting pointer

        for (int right = 0; right < arr.length; right++) {
            currentSum += arr[right]; // Window ko right se badao

            // Agar hamari current window ka sum abhi tak ka sabse chota hai
            if (currentSum < minSum) {
                minSum = currentSum;
            }

            // Sliding Window Rule: Agar window ka total positive (> 0) ho gaya,
            // toh yeh aage ke sum ko bada karega, chota nahi. Isliye window shrink/reset karo.
            if (currentSum > 0) {
                currentSum = 0;
                left = right + 1; // Window ko naye element par shift karo
            }
        }
        return minSum;
    }
}

File: Difficulty: Medium/Minimum Sum Subarray/minimum-sum-subarray.cpp
class Solution {
  public:
    int minSubarraySum(vector<int> &arr) {
        // code here
        // kadens se hua hai
        int min_sum = arr[0], current_sum = 0 , n = arr.size();
        
        for(int i = 0;i < n;i++){
            current_sum += arr[i];
            
            if(current_sum < min_sum){
                min_sum = current_sum;
            }
            if (current_sum > 0){
                current_sum = 0;
            }
        }
        return min_sum;
        
        // sliding window ka 
    }
};